File: src/main/java/kh/farrukh/aoc2024/day1/Day1B.java
package kh.farrukh.aoc2024.day1;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;

public class Day1B {

  public static void main(String[] args) throws IOException {
    BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

    var list1 = new ArrayList<Integer>();
    var list2 = new ArrayList<Integer>();

    var l = reader.readLine();

    while (l != null && !l.isEmpty()) {
      var ns = l.split("   ");
      list1.add(Integer.parseInt(ns[0]));
      list2.add(Integer.parseInt(ns[1]));
      l = reader.readLine();
    }

    var similarity = new HashMap<Integer, Integer>();

    for (int i = 0; i < list1.size(); i++) {
      var n1 = list1.get(i);
      var simCount = list2.stream().filter(x -> x.equals(n1)).count();
      similarity.put(n1, (int) simCount);
    }

    var res = 0;

    for (var entry : similarity.entrySet()) {
      res += entry.getKey() * entry.getValue();
    }

    System.out.println(res);
  }

}

File: src/main/java/kh/farrukh/aoc2024/day1/Day1B.kt
package kh.farrukh.aoc2024.day1

import java.util.Scanner
import java.util.stream.Collectors
import kotlin.math.abs


fun main() {
    val sc = Scanner(System.`in`)

    val list1 = mutableListOf<Int>()
    val list2 = mutableListOf<Int>()

    var line = sc.nextLine()

    while (line != null && line.isNotBlank()) {
        val nums = line.split("   ")
        list1.add(nums[0].toInt())
        list2.add(nums[1].toInt())

        line = sc.nextLine()
    }

    val list2Map = list2.groupingBy { it }.eachCount()

    var res = 0

    list1.forEach {
        list2Map[it]?.let { count -> res += it * count }
    }

    println(res)
}
